ql/src/java/org/apache/hadoop/hive/ql/exec/repl/bootstrap/load/LoadConstraint.java

```diff
@@ -0,0 +1,119 @@
+import org.apache.hadoop.hive.ql.parse.repl.DumpType;
+import org.apache.hadoop.hive.ql.parse.repl.load.DumpMetaData;
+import org.apache.hadoop.hive.ql.parse.repl.load.message.AddForeignKeyHandler;
+import org.apache.hadoop.hive.ql.parse.repl.load.message.AddNotNullConstraintHandler;
+import org.apache.hadoop.hive.ql.parse.repl.load.message.AddPrimaryKeyHandler;
+import org.apache.hadoop.hive.ql.parse.repl.load.message.AddUniqueConstraintHandler;
+import org.apache.hadoop.hive.ql.parse.repl.load.message.MessageHandler;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.apache.hadoop.hive.ql.parse.BaseSemanticAnalyzer.stripQuotes;
+
+public class LoadConstraint {
+  private static final Logger LOG = LoggerFactory.getLogger(LoadFunction.class);
+  private Context context;
+  private final ConstraintEvent event;
+  private final String dbNameToLoadIn;
+  private final TaskTracker tracker;
+
+  public LoadConstraint(Context context, ConstraintEvent event, String dbNameToLoadIn,
+      TaskTracker existingTracker) {
+    this.context = context;
+    this.event = event;
+    this.dbNameToLoadIn = dbNameToLoadIn;
+    this.tracker = new TaskTracker(existingTracker);
+  }
+
+  public TaskTracker tasks() throws IOException, SemanticException {
+    URI fromURI = EximUtil
+        .getValidatedURI(context.hiveConf, stripQuotes(event.rootDir().toUri().toString()));
+    Path fromPath = new Path(fromURI.getScheme(), fromURI.getAuthority(), fromURI.getPath());
+
+    try {
+      FileSystem fs = FileSystem.get(fromPath.toUri(), context.hiveConf);
+      JSONObject json = new JSONObject(EximUtil.readAsString(fs, fromPath));
+      String pksString = json.getString("pks");
+      String fksString = json.getString("fks");
+      String uksString = json.getString("uks");
+      String nnsString = json.getString("nns");
+      List<Task<? extends Serializable>> tasks = new ArrayList<Task<? extends Serializable>>();
+
+      AddPrimaryKeyHandler pkHandler = new AddPrimaryKeyHandler();
+      DumpMetaData pkDumpMetaData = new DumpMetaData(fromPath, DumpType.EVENT_ADD_PRIMARYKEY, Long.MAX_VALUE, Long.MAX_VALUE, null,
+          context.hiveConf);
+      pkDumpMetaData.setPayload(pksString);
+      tasks.addAll(pkHandler.handle(
+          new MessageHandler.Context(
+              dbNameToLoadIn, null, fromPath.toString(), null, pkDumpMetaData, context.hiveConf,
+              context.hiveDb, null, LOG)));
+
+      AddForeignKeyHandler fkHandler = new AddForeignKeyHandler();
+      DumpMetaData fkDumpMetaData = new DumpMetaData(fromPath, DumpType.EVENT_ADD_FOREIGNKEY, Long.MAX_VALUE, Long.MAX_VALUE, null,
+          context.hiveConf);
+      fkDumpMetaData.setPayload(fksString);
+      tasks.addAll(fkHandler.handle(
+          new MessageHandler.Context(
```

Comment:
In DDLTask, we do check repl ID for all alterTable, including addConstaint:
      if (alterTbl != null) {
        if (!allowOperationInReplicationScope(db, alterTbl.getOldName(), null, alterTbl.getReplicationSpec())) {
          // no alter, the table is missing either due to drop/rename which follows the alter.
          // or the existing table is newer than our update.
          LOG.debug("DDLTask: Alter Table is skipped as table {} is newer than update", alterTbl.getOldName());
          return 0;
        }
        if (alterTbl.getOp() == AlterTableTypes.DROPCONSTRAINT ) {
          return dropConstraint(db, alterTbl);
        } else if (alterTbl.getOp() == AlterTableTypes.ADDCONSTRAINT) {
          return addConstraints(db, alterTbl);
        } else {
          return alterTable(db, alterTbl);
        }
      }

Reply:
The problem is the table object was captured long before constraints are captured. So, the last repl ID stored in table properties might be outdated with respect to constraints. For instance, if a constraint is added after dumping table object and before dumping constraints, then newly added constraint will be part of the bootstrap dump and will also be applied as part of catch-up incremental repl. During catch-up incremental repl, it will be tried to create constraint again as last repl id is older than the create_constraint event itself. So, need to handle this case.

Reply:
I see. We need to ignore already exist exception when doing incremental constraint loading. May I create a separate ticket for that? As I'd like to check this in sooner.

Reply:
Sounds good.